File: app/tasks/stream_runner.py
import asyncio
import traceback

from app.core.logger import logger

from app.websockets.market_stream import (
    MarketStream
)


async def main():

    stream = MarketStream()

    await stream.start()


async def run_with_recovery():
    while True:
        try:
            await main()
        except Exception as e:
            logger.error(f"Stream Runner crashed: {e}")
            logger.error(traceback.format_exc())
            logger.info("Restarting Stream Runner in 5 seconds...")
            await asyncio.sleep(5)


if __name__ == "__main__":

    asyncio.run(run_with_recovery())

File: backend/src/workers/streamRunner.ts
import { MarketStream } from "../websockets/marketStream";
import { logger } from "../core/logger";

async function main() {
  const stream = new MarketStream();
  await stream.start();
}

async function runWithRecovery() {
  while (true) {
    try {
      await main();
      // Keep process alive since ws client is event-driven
      await new Promise(() => {});
    } catch (e: any) {
      logger.error(`Stream Runner crashed: ${e.message}`);
      if (e.stack) logger.error(e.stack);
      logger.info("Restarting Stream Runner in 5 seconds...");
      await new Promise((resolve) => setTimeout(resolve, 5000));
    }
  }
}

runWithRecovery();
